airsonic-main/src/main/java/org/airsonic/player/dao/LegacyHsqlDaoHelper.java

```diff
@@ -31,29 +51,226 @@ public void checkpoint() {
+            return false;
+        } else {
+            // If this happens we're on a completely untested version and we don't know what will happen.
+            LOG.warn("HSQLDB database upgrade from version {} to version {} is untested. Trying to connect anyway, this may upgrade the database.", currentVersion, driverVersion);
+            return false;
+        }
+    }
+
+    /**
+     * Perform a backup of the HSQLDB database, to a timestamped directory.
+     * @return the path to the backup directory
+     */
+    private Path performHsqldbDatabaseBackup() throws IOException {
+
+        String timestamp = new SimpleDateFormat("yyyyMMddHHmmss").format(new Date());
+        Path source = Paths.get(settingsService.getDefaultJDBCPath()).getParent();
+        Path destination = Paths.get(String.format("%s.backup.%s", settingsService.getDefaultJDBCPath(), timestamp));
+
+        LOG.info("Backing up HSQLDB database to {}...", destination.toString());
+        try (Stream<Path> stream = Files.walk(source)) {
+
+            List<Path> sources = stream.collect(toList());
+            List<Path> destinations = sources.stream()
+                    .map(source::relativize)
+                    .map(destination::resolve)
+                    .collect(toList());
+
+            for (int i = 0; i < sources.size(); i++) {
+                LOG.info("Copying {} to {}", sources.get(i), destinations.get(i));
+                Path currentSource = sources.get(i);
+                Path currentDestination = destinations.get(i);
+                if (currentDestination.toFile().isDirectory()) {
+                    currentDestination.toFile().mkdirs();
+                } else {
+                    currentDestination.getParent().toFile().mkdirs();
+                    Files.copy(sources.get(i), destinations.get(i));
+                }
+            }
+        }
+
+        LOG.info("HSQLDB database backup complete.");
+        return destination;
+    }
+
+    /**
+     * Perform an in-place database upgrade from HSQLDB 1.x to 2.x.
+     */
+    private void performHsqldbDatabaseUpgrade() throws SQLException {
+        // This will upgrade HSQLDB from 1.8 to 2.5 on the first connection.
+        try (Connection conn = DriverManager.getConnection(settingsService.getDefaultJDBCUrl())) {
+            LOG.info("Database connection established. Current version is: {}.", conn.getMetaData().getDatabaseProductVersion());
+            // On upgrade, the official documentation recommends that we
+            // run 'SHUTDOWN SCRIPT' to compact all the database into a
+            // single SQL file.
+            //
+            // In practice, if we don't do that, we did not observe issues
+            // immediately but after the upgrade.
+            LOG.info("Shutting down database...");
+            try (Statement st = conn.createStatement()) {
+                st.execute("SHUTDOWN SCRIPT");
+            }
```

Comment:
shouldn't we just use the datasource jdbctemplate injected into this class?
```
getJdbcTemplate().execute("SHUTDOWN SCRIPT");
```

Reply:
I would actually prefer to add another statement immediately afterwards to check if the connection is reestablished after shutdown

Reply:
We cannot actually use the `DataSource` that Airsonic is going to use after the upgrade, because it will be closed. I did not find any way to restore the connection after a `SHUTDOWN`, do you have an idea?

Reply:
Hmmm, then how do we expect the application to start up after the shutdown? Do people have to start the application twice?
One idea is to shutdown the connection pool and then restart it...